### PR TITLE
Add Datadog::Shim for instrumentation hooks

### PR DESCRIPTION
**Pull request by delner**
To expand our toolbox for building integrations, this pull request introduces the `Datadog::Shim` module. This module introduces behaviors for adding hooks around existing code, with the intent that it should be used to add instrumentation hooks to libraries.

There are a few components added here:

 - `Wrapping` module: a module that can be composed into a class, which adds methods for wrapping methods on that object via `#inject_method!` and `#wrap_method!`.
 - `Double` class: This is a "man-in-the-middle" object, that can wrap around any other object, and intercept messages sent to it, while forwarding the rest to the original object. It is useful when you want to modify behavior without modifying the original object or its class. (e.g. this could be applicable if you want an object to behave differently when called from a specific source, rather than changing the behavior in a global way.)
 - `MethodWrapper` class: defines a wrapper for a method which can be invoked. Similar to RSpec's `around(:each) { |example| }` method.

Things left to do:

- [ ] Add `Wrapping` module
- [ ] Add ``Double` class
- [ ] Performance tests: compare to non-wrapped function
- [ ] Add specs

This is an experimental feature: if we want this, great, otherwise, we can close this PR.

Currently being leveraged in [an alternative implementation of instrumenting ActiveRecord connection properties](https://github.com/DataDog/dd-trace-rb/compare/fix/active_record_avoid_connection_id2ref?expand=1). See #640 for more context.

### CODE REVIEW

**Comment by delner** [comment]
Did some performance tests for this. My main concern is that the use of such `Double` or wrapper methods is that the performance cost is inexpensive. Through experimentation, the general rule of thumb seemed to be `the fewer blocks, the faster it is`. The implementation I have currently tries to minimize the number of `#call` or `#instance_eval` calls to keep performance as tight as possible.

In the most bare minimum example, given a class with method `def foo; :foo end`, in which `foo` is called over apx. 1M iterations, here's how the performance roughly breaks down.

```
                                         user     system      total        real
control                              0.330000   0.000000   0.330000 (  0.332207)
Double#wrap_method! with block       0.640000   0.000000   0.640000 (  0.644946)
Double#wrap_method! without args     0.590000   0.000000   0.590000 (  0.587711)
Double#inject_method! with block     0.510000   0.010000   0.520000 (  0.528495)
Double#inject_method! without args   0.450000   0.000000   0.450000 (  0.454361)
```

Basically the "worst case" takes roughly double the time vs the control. However this is a comparison of overhead vs `def foo; :foo end`, which really is the lightest function you could possibly have, so any tiny performance hit we introduce will be exaggerated in this test. I would expect that in real applications, most functions worthy of being instrumented will be far more expensive, so the overhead this instrumentation adds will become a very small percentile overall. In that context, I think this is pretty reasonable.

There are some other experiments that would be worth trying, like `def method_missing` or other alternative implementations, which might be faster/better.

File: lib/ddtrace/shim.rb [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        # Forward methods
+        forwarded_methods = shim_target.public_methods - EXCLUDED_METHODS - wrapped_methods.to_a
+        forward_methods!(*forwarded_methods)
+      end
+
+      def forward_methods!(*forwarded_methods)
+        return if forwarded_methods.empty?
+
+        singleton_class.send(
+          :def_delegators,
```
Comment (pawelchcki) [inline lib/ddtrace/shim.rb]: I tried looking up how `:def_delegators` is implemented. My concern is what is the performance impact of instrumenting object with many, many methods.

**Comment by pawelchcki** [comment]
Awesome stuff! Just one question about how the performance changes based on number of methods in instrumented Object.

File: lib/ddtrace/shim.rb (hunk 1 above)
Comment (delner) [reply lib/ddtrace/shim.rb]: As in if there are more arugments provided to that function than less? I wouldn't have imagined the number of arguments would matter here, but its a good question. I don't know how the performance cost scales, or how it compares to others, spare what I saw in https://gist.github.com/rbishop/7555357.

**Comment by delner** [comment]
Okay, I'll add some specs to cover the new components, then I think this'll be ready to merge.
